python/lsst/ip/isr/calibType.py

```diff
@@ -205,6 +218,63 @@ def updateMetadata(self, camera=None, detector=None, filterName=None,
         mdSupplemental.update(kwargs)
         mdOriginal.update(mdSupplemental)
 
+    def calibInfoFromDict(self, dictionary):
+        """Handle common keywords.
+
+        This isn't an ideal solution, but until all calibrations
+        expect to find everything in the metadata, they still need to
+        search through dictionaries.
+
+        Parameters
+        ----------
+        dictionary : `dict` or `lsst.daf.base.PropertyList`
+            Source for the common keywords.
+
+        Raises
+        ------
+        RuntimeError :
+            Raised if the dictionary does not match the expected OBSTYPE.
+
+        """
+
+        def search(haystack, needles):
+            """Search dictionary 'haystack' for an entry in 'needles'
+            """
+            test = [haystack.get(x) for x in needles]
+            test = set([x for x in test if x is not None])
+            if len(test) == 0:
+                if 'metadata' in haystack:
+                    return search(haystack['metadata'], needles)
+                else:
+                    return None
+            elif len(test) == 1:
+                value = list(test)[0]
+                if value == '':
+                    return None
+                else:
+                    return value
+            else:
+                raise ValueError(f"Too many values found: {len(test)} {test} {needles}")
+
+        if 'metadata' in dictionary:
+            metadata = dictionary['metadata']
+
+            if self._OBSTYPE != metadata['OBSTYPE']:
+                raise RuntimeError(f"Incorrect calibration supplied.  Expected {self._OBSTYPE}, "
+                                   f"found {metadata['OBSTYPE']}")
+
+        self._instrument = search(dictionary, ['INSTRUME', 'instrument'])
+        self._raftName = search(dictionary, ['RAFTNAME'])
+        self._slotName = search(dictionary, ['SLOTNAME'])
+        self._detectorId = search(dictionary, ['DETECTOR', 'detectorId'])
+        self._detectorName = search(dictionary, ['DET_NAME', 'DETECTOR_NAME', 'detectorName'])
+        self._detectorSerial = search(dictionary, ['DET_SER', 'DETECTOR_SERIAL', 'detectorSerial'])
+        self._filter = search(dictionary, ['FILTER', 'filterName'])
+        self._calibId = search(dictionary, ['CALIB_ID'])
+
+        if not self._detectorId and self._detectorSerial:
+            self._detectorId = self._detectorSerial
```

Comment:
If you have no detId but you do have a serial, just make the detId equal to the serial? 🤨 I don't really understand how these can be interchangeable - is there scary downstream logic somewhere/everywhere for dealing with this, or am I misunderstanding?

Reply:
I'm removing this, as I think all of the cases that were inconsistent with the detectorId have been purged from the code.